src/main/java/com/zepben/evolve/cim/iec61970/base/wires/PowerTransformer.kt

```diff
@@ -53,7 +54,12 @@ class PowerTransformer @JvmOverloads constructor(mRID: String = "") : Conducting
 
     private var _powerTransformerEnds: MutableList<PowerTransformerEnd>? = null
     var vectorGroup: VectorGroup = VectorGroup.UNKNOWN
-    var transformerUtilisation: Double = 0.0
+    var transformerUtilisation: Double = Double.NaN
```

Comment:
There is also the memory overhead of a nullable primitive type to consider. 
Another thing is for all our other primitives we don't use nullables for "we don't have this data". We just use a default value, though I'm not sure if they are potential real default values or they are rubbish default values. 0 was a valid value in this case and it was causing problems for Geordie.

Reply:
Yep, originally I made it null but 

> What are we actually adding powerTransformerInfo for if it has no fields?

Essentially, just to hold an mrid. Geordie has access to a microsoft access db that holds all the transformer datasheet information. So he wants to know the id of the datasheet that is relevant for each transformer. He needs this to run his sincal models.
We have that id available in the geojson payload that we get from essential so the plan is to use it as the id for the powerTransformerInfo.

Reply:
???? why did this appear here, I would have expected it to appear as a reply to the "What are we actually adding powerTransformerInfo for if it has no fields?" comment

Reply:
> is this to cover the "we don't have this data" case? I think it'd make more sense to make it nullable instead, that way you at least get the kotlin null safety benefits. The overhead is obviously converting from null to NaN when serialising to/from protobuf, but I don't think that's a big deal.
> There is also the memory overhead of a nullable primitive type to consider.
> Another thing is for all our other primitives we don't use nullables for "we don't have this data". We just use a default value, though I'm not sure if they are potential real default values or they are rubbish default values. 0 was a valid value in this case and it was causing problems for Geordie.

I can see arguments for both, I am leaning a bit towards leaving the NaN if we don't use nullables for "we don't have this data" throughout the codebase just for the sake of consistency but I certainly see Kurts point of taking advantage of kotlin's null safety because someone just making use of that value for some operation expecting it to be a regular number can get burned by the fact they didn't expect a NaN.

Even though NaN, -Infinity, and +Infinity are valid floating point values I rarely see code that handles those cases because they are so rare.

So we just need to make a decision and stick with it for now I guess. Do you guys want me to create a meeting on zoom or something so we can discuss the pros and cons of each and make a decision?

Reply:
> though I'm not sure if they are potential real default values or they are rubbish default values. 0 was a valid value in this case and it was causing problems for Geordie.

for most of them they are obviously rubbish I believe, however we'll likely run into more cases where this problem crops up. this will only occur for number based types and using NaN raises the issue that how we handle this problem will be inconsistent between integers and doubles (as Integers don't have NaN and thus will have to be nullable). Seems to me like that could be undesirable as you'd potentially write calculations that utilise both integers and doubles, and doing a bunch of NaN checks and a bunch of null checks would be confusing.

in this case there'll never be so many PowerTransformers that the overhead of a nullable type will be significant I'd think, but if it sets a precedent it's possibly problematic for other types.

Reply:
Yes, just for this one field it would not be significant. But if we haven't done it for anywhere else already, and for those fields you already have to check for a rubbish value then I think being consistent here with a primitive with a  rubbish value for now makes sense.

Changing all the primitives to nullable across the board needs to be assessed for real world memory / performance impact. It is also a breaking change for Kotlin users so the flow through effects could be large.